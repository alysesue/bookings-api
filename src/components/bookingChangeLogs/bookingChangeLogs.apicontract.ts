import { DynamicValueContract } from '../dynamicFields/dynamicValues.apicontract';
import { UserTypeResponse } from '../users/users.apicontract';

export class BookingChangeLogResponseV1 {
	/**
	 * @isInt
	 */
	public bookingId: number;
	public changeLogs: ChangeLogEntryResponseV1[];
}

export class BookingChangeLogResponseV2 {
	public bookingId: string;
	public changeLogs: ChangeLogEntryResponseV2[];
}

export class ChangeLogEntryResponseBase {
	public timestamp: Date;
	public user: UserTypeResponse;
	public action: ChangeLogActionContract;
}

export class ChangeLogEntryResponseV1 extends ChangeLogEntryResponseBase {
	public previousBooking: BookingStateResponseV1;
	public changes: BookingStateResponseV1;
}

export class ChangeLogEntryResponseV2 extends ChangeLogEntryResponseBase {
	public previousBooking: BookingStateResponseV2;
	public changes: BookingStateResponseV2;
}

export class BookingStateResponseBase {
	/**
	 * @isInt
	 */
	public status?: number;
	public startDateTime?: Date;
	public endDateTime?: Date;
	public serviceName?: string;
	public serviceProviderAgencyUserId?: string;
	public serviceProviderName?: string;
	public serviceProviderEmail?: string;
	public serviceProviderPhone?: string;
	public citizenUinFin?: string;
	public citizenName?: string;
	public citizenEmail?: string;
	public citizenPhone?: string;
	public location?: string;
	public description?: string;
	public videoConferenceUrl?: string;
	public refId?: string;
	public dynamicValues?: DynamicValueContract[];
	public serviceProviderAliasName?: string;
}

export class BookingStateResponseV1 extends BookingStateResponseBase {
	/**
	 * @isInt
	 */
	public id?: number;
	/**
	 * @isInt
	 */
	public serviceId?: number;
	/**
	 * @isInt
	 */
	public serviceProviderId?: number;
}

export class BookingStateResponseV2 extends BookingStateResponseBase {
	public id?: string;
	public serviceId?: string;
	public serviceProviderId?: string;
}

export type ChangeLogSearchRequest = {};

export enum ChangeLogActionContract {
	Create = 'create',
	Accept = 'accept',
	Reject = 'reject',
	Cancel = 'cancel',
	Update = 'update',
	Reschedule = 'reschedule',
	UpdateUser = 'update-user',
}
