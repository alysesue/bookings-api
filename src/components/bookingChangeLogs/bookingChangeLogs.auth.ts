import { QueryAuthGroupVisitor } from '../../infrastructure/auth/queryAuthGroupVisitor';
import {
	AuthGroup,
	CitizenAuthGroup,
	OrganisationAdminAuthGroup,
	ServiceAdminAuthGroup,
	ServiceProviderAuthGroup,
} from '../../infrastructure/auth/authGroup';
import { UserConditionParams } from '../../infrastructure/auth/authConditionCollection';

export class BookingChangeLogsQueryAuthVisitor extends QueryAuthGroupVisitor {
	private readonly _alias: string;

	constructor(alias: string) {
		super();
		this._alias = alias;
	}

	public visitCitizen(_citizenGroup: CitizenAuthGroup): void {}

	public visitOrganisationAdmin(_userGroup: OrganisationAdminAuthGroup): void {
		const authorisedOrganisationIds = _userGroup.authorisedOrganisations.map((org) => org.id);
		this.addAuthCondition(`${this._alias}."_organisationId" IN (:...authorisedOrganisationIds)`, {
			authorisedOrganisationIds,
		});
	}

	public visitServiceAdmin(_userGroup: ServiceAdminAuthGroup): void {
		const authorisedServiceIds = _userGroup.authorisedServices.map((s) => s.id);
		this.addAuthCondition(`${this._alias}."_serviceId" IN (:...authorisedServiceIds)`, {
			authorisedServiceIds,
		});
	}

	public visitServiceProvider(_userGroup: ServiceProviderAuthGroup): void {
		const authorisedServiceProviderIds = _userGroup.authorisedServiceProvider.id;
		this.addAuthCondition(`${this._alias}."serviceProviderId" = :authorisedServiceProviderIds`, {
			authorisedServiceProviderIds,
		});
	}
}

class BookingChangeLogsQueryNoAuthVisitor extends BookingChangeLogsQueryAuthVisitor {
	public async createUserVisibilityCondition(authGroups: AuthGroup[]): Promise<UserConditionParams> {
		this.addAsTrue();
		return this.getVisibilityCondition();
	}
}

export class BookingChangeLogQueryVisitorFactory {
	public static getBookingChangeLogQueryVisitor(byPassAuth: boolean): BookingChangeLogsQueryAuthVisitor {
		if (byPassAuth) {
			return new BookingChangeLogsQueryNoAuthVisitor('bookingChangeLog');
		}
		return new BookingChangeLogsQueryAuthVisitor('bookingChangeLog');
	}
}
