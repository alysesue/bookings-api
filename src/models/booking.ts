import {BaseEntity, Column, Entity, PrimaryGeneratedColumn} from "typeorm";
import {BookingStatus} from "./bookingStatus";

@Entity()
export class Booking extends BaseEntity {
	@PrimaryGeneratedColumn()
	private _id: number;

	@Column()
	private _status: BookingStatus;

	@Column()
	private _startDateTime: Date;
	@Column()
	private _sessionDurationInMinutes: number;

	constructor(startDateTime: Date, sessionDurationInMinutes: number) {
		super();
		this._startDateTime = startDateTime;
		this._sessionDurationInMinutes = sessionDurationInMinutes;
		this._status = BookingStatus.PendingApproval;
	}

	public get id(): number {
		return this._id;
	}

	public set status(newStatus: BookingStatus) {
		this._status = newStatus;
	}

	public get status(): BookingStatus {
		return this._status;
	}

	public get startDateTime(): Date {
		return this._startDateTime;
	}

	public get sessionDurationInMinutes(): number {
		return this._sessionDurationInMinutes;
	}

	public getSessionEndTime(): Date {
		return new Date(this._startDateTime.getTime() + this._sessionDurationInMinutes * 60 * 1000);
	}
}
