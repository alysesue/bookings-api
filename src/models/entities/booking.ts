import { Column, Entity, Index, JoinColumn, ManyToOne, PrimaryGeneratedColumn } from "typeorm";
import { BookingStatus } from "../bookingStatus";
import { ServiceProvider } from './serviceProvider';
import { Service } from "./service";
import * as timeSpan from "../../tools/timeSpan";
import { User } from "./user";

@Entity()
export class Booking {

	@PrimaryGeneratedColumn()
	private _id: number;

	@Column({ nullable: false })
	private _serviceId: number;

	@ManyToOne(type => Service)
	@JoinColumn({ name: '_serviceId' })
	private _service: Service;

	@Column({ type: "varchar", length: 300, nullable: true })
	private _eventICalId: string;

	@Column()
	private _status: BookingStatus;

	@Column()
	@Index()
	private _startDateTime: Date;

	@Column()
	@Index()
	private _endDateTime: Date;

	@Column()
	private _createdAt: Date;

	@Column({ nullable: true })
	private _refId?: string;

	@Column({ nullable: true })
	private _acceptedAt: Date;

	@ManyToOne(type => ServiceProvider, { nullable: true })
	@JoinColumn({ name: '_serviceProviderId' })
	private _serviceProvider: ServiceProvider;

	@Column({ nullable: true })
	private _serviceProviderId?: number;

	@ManyToOne(type => User, { cascade: true, nullable: true })
	@JoinColumn({ name: '_citizenUserId' })
	private _citizenUser: User;

	@Column({ nullable: true })
	private _citizenUserId: number;

	public get citizenUserId(): number {
		return this._citizenUserId;
	}

	public set citizenUserId(value: number) {
		this._citizenUserId = value;
	}

	constructor() {
	}

	public get id(): number {
		return this._id;
	}

	public get serviceId(): number {
		return this._serviceId;
	}

	public get service(): Service {
		return this._service;
	}

	public get eventICalId(): string {
		return this._eventICalId;
	}

	public set eventICalId(value: string) {
		this._eventICalId = value;
	}

	public get status(): BookingStatus {
		return this._status;
	}

	public set status(newStatus: BookingStatus) {
		this._status = newStatus;
	}

	public get startDateTime(): Date {
		return this._startDateTime;
	}

	public get endDateTime(): Date {
		return this._endDateTime;
	}

	public set acceptedAt(acceptedAt: Date) {
		this._acceptedAt = acceptedAt;
	}

	public get serviceProvider(): ServiceProvider {
		return this._serviceProvider;
	}

	public set serviceProvider(serviceProvider: ServiceProvider) {
		this._serviceProvider = serviceProvider;
	}

	public set serviceProviderId(value: number | undefined) {
		this._serviceProviderId = value;
	}

	public get serviceProviderId(): number | undefined {
		return this._serviceProviderId;
	}

	public get citizenUser(): User {
		return this._citizenUser;
	}

	public set citizenUser(value: User) {
		this._citizenUser = value;
	}

	public get createdAt(): Date {
		return this._createdAt;
	}

	public bookingIntersects(other: { start: Date, end: Date }): boolean {
		if (!other.start || !other.end) {
			return false;
		}
		return timeSpan.intersectsDateTimeSpan(other, this.startDateTime, this.endDateTime);
	}

	public static create(serviceId: number, startDateTime: Date, endDateTime: Date, serviceProviderId?: number, refId?: string, eventICalId?: string) {
		const instance = new Booking();
		instance._serviceId = serviceId;
		instance._startDateTime = startDateTime;
		instance._endDateTime = endDateTime;
		instance._createdAt = new Date();
		instance._refId = refId;
		instance._eventICalId = eventICalId;

		if (serviceProviderId) {
			instance._serviceProviderId = serviceProviderId;
			instance._status = BookingStatus.Accepted;
			instance._acceptedAt = instance.createdAt;
		} else {
			instance._status = BookingStatus.PendingApproval;
		}

		return instance;
	}
}
