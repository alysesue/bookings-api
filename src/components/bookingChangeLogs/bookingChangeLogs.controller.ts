import { Inject } from 'typescript-ioc';
import { Controller, Get, Header, Query, Response, Route, Security, SuccessResponse, Tags } from 'tsoa';
import { MOLAuth } from 'mol-lib-common';
import { MOLUserAuthLevel } from 'mol-lib-api-contract/auth/auth-forwarder/common/MOLUserAuthLevel';
import { BookingChangeLogResponse } from './bookingChangeLogs.apicontract';
import { BookingChangeLogsService } from './bookingChangeLogs.service';
import { BookingChangeLogsMapper } from './bookingChangeLogs.mapper';

@Route('v1/bookinglogs')
@Tags('BookingLogs')
export class BookingChangeLogsController extends Controller {
	@Inject
	private changeLogsService: BookingChangeLogsService;

	/**
	 * Retrieves all booking logs in the specified datetime range [changedSince, changedUntil).
	 * @param changedSince The lower bound datetime limit (inclusive) for logs' timestamp.
	 * @param changedUntil The upper bound datetime limit (exclusive) for logs' timestamp.
	 * @param bookingIds (Optional) filters by a list of booking ids.
	 * @param serviceId (Optional) filters by a service (id).
	 */
	@Get('')
	@SuccessResponse(200, 'Ok')
	@Security('optional-service')
	@MOLAuth({
		admin: {},
		user: { minLevel: MOLUserAuthLevel.L2 },
	})
	@Response(401, 'Valid authentication types: [admin,user]')
	public async getChangeLogs(
		@Query() changedSince: Date,
		@Query() changedUntil: Date,
		@Query() bookingIds?: number[],
		@Header('x-api-service') serviceId?: number,
	): Promise<BookingChangeLogResponse[]> {
		const logs = await this.changeLogsService.getLogs({ changedSince, changedUntil, serviceId, bookingIds });
		return BookingChangeLogsMapper.mapDataModels(logs);
	}
}
