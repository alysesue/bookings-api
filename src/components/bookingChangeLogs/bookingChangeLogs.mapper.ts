import { Inject, InRequestScope } from 'typescript-ioc';
import { BookingChangeLog, BookingJsonSchemaV1, ChangeLogAction } from '../../models';
import { UserProfileMapper } from '../users/users.mapper';
import {
	BookingChangeLogResponse,
	BookingStateResponse,
	ChangeLogActionContract,
	ChangeLogEntryResponse,
} from './bookingChangeLogs.apicontract';
import { DynamicValuesMapper } from '../dynamicFields/dynamicValues.mapper';
import * as _ from 'lodash';

type GenericState = { [key: string]: any };

@InRequestScope
export class BookingChangeLogsMapper {
	private static _actionMap: Readonly<{ [key: number]: ChangeLogActionContract }>;

	@Inject
	private dynamicValuesMapper: DynamicValuesMapper;

	private constructor() {}

	public static initialize() {
		const map: { [key: number]: ChangeLogActionContract } = {};
		map[ChangeLogAction.Create] = ChangeLogActionContract.Create;
		map[ChangeLogAction.Accept] = ChangeLogActionContract.Accept;
		map[ChangeLogAction.Reject] = ChangeLogActionContract.Reject;
		map[ChangeLogAction.Cancel] = ChangeLogActionContract.Cancel;
		map[ChangeLogAction.Update] = ChangeLogActionContract.Update;
		map[ChangeLogAction.Reschedule] = ChangeLogActionContract.Reschedule;

		BookingChangeLogsMapper._actionMap = map;
	}

	private mapChangeLogAction(action: ChangeLogAction): ChangeLogActionContract {
		const value = BookingChangeLogsMapper._actionMap[action];
		if (!value) {
			throw new Error('No mapping found for ChangeLogAction: ' + action);
		}
		return value;
	}

	private mapBookingState(state: BookingJsonSchemaV1): BookingStateResponse {
		const { dynamicValues, ...fields } = state;

		const mappedDynamicValues = this.dynamicValuesMapper.mapDynamicValuesModel(dynamicValues);
		return { ...fields, dynamicValues: mappedDynamicValues };
	}

	private getChanges<T extends GenericState>(previousState: T, newState: T): T {
		const changes: GenericState = {};
		for (const key of Object.keys(newState)) {
			if (!_.isEqual(newState[key], previousState[key])) {
				changes[key] = newState[key];
			}
		}
		return changes as T;
	}

	public mapChangeLog(changeLog: BookingChangeLog): ChangeLogEntryResponse {
		const instance = new ChangeLogEntryResponse();
		instance.timestamp = changeLog.timestamp;
		instance.user = UserProfileMapper.mapUserToResponse(changeLog.user);
		instance.action = this.mapChangeLogAction(changeLog.action);
		instance.previousBooking = this.mapBookingState(changeLog.previousState);
		const changes = this.getChanges(changeLog.previousState, changeLog.newState);
		instance.changes = this.mapBookingState(changes);

		return instance;
	}

	public mapDataModels(changeLogs: Map<number, BookingChangeLog[]>): BookingChangeLogResponse[] {
		const result: BookingChangeLogResponse[] = [];
		for (const [bookingId, logs] of changeLogs.entries()) {
			const bookingEntry = new BookingChangeLogResponse();
			bookingEntry.bookingId = bookingId;
			bookingEntry.changeLogs = logs.map((log) => this.mapChangeLog(log));
			result.push(bookingEntry);
		}

		return result;
	}
}

BookingChangeLogsMapper.initialize();
