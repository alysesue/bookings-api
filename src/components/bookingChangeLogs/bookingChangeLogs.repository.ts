import { Inject, InRequestScope } from 'typescript-ioc';
import { BookingChangeLog } from '../../models';
import { RepositoryBase } from '../../core/repository';
import { UserContext } from '../../infrastructure/userContext.middleware';
import { groupByKey } from '../../tools/collections';

@InRequestScope
export class BookingChangeLogsRepository extends RepositoryBase<BookingChangeLog> {
	@Inject
	private userContext: UserContext;

	constructor() {
		super(BookingChangeLog);
	}

	public async save(changeLog: BookingChangeLog): Promise<BookingChangeLog> {
		const repository = await this.getRepository();
		return repository.save(changeLog);
	}

	public async getLogs(options: ChangeLogSearchQuery): Promise<Map<number, BookingChangeLog[]>> {
		const { changedSince, changedUntil, serviceId, bookingIds } = options;

		const serviceCondition = serviceId ? 'changelog."_serviceId" = :serviceId' : '';
		const dateCondition = 'changelog."_timestamp" >= :changedSince AND changelog."_timestamp" < :changedUntil';
		const bookingIdsCondition = bookingIds ? 'changelog."_bookingId" IN (:...bookingIds)' : '';

		const repository = await this.getRepository();
		const query = repository.createQueryBuilder('changelog').where(
			[serviceCondition, dateCondition, bookingIdsCondition]
				.filter((c) => c)
				.map((c) => `(${c})`)
				.join(' AND '),
			{ changedSince, changedUntil, serviceId, bookingIds },
		);

		const entries = await query.getMany();

		return groupByKey(entries, (log) => log.bookingId);
	}
}

export type ChangeLogSearchQuery = {
	changedSince: Date;
	changedUntil: Date;
	serviceId?: number;
	bookingIds: number[];
};
