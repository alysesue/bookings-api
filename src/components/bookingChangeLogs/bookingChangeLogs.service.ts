import { Inject, InRequestScope } from "typescript-ioc";
import { Booking } from "../../models";
import { UserContext } from "../../infrastructure/userContext.middleware";
import { BookingChangeLog, BookingJsonSchemaV1, ChangeLogAction } from "../../models/entities/bookingChangeLog";
import { BookingChangeLogsRepository } from "./bookingChangeLogs.repository";
import { TransactionManager } from "../../core/transactionManager";
import { IsolationLevel } from "typeorm/driver/types/IsolationLevel";

export type BookingActionFunction = (booking: Booking) => Promise<[ChangeLogAction, Booking]>;

@InRequestScope
export class BookingChangeLogsService {
	private static readonly isolationLevel: IsolationLevel = "READ COMMITTED";

	@Inject
	private transactionManager: TransactionManager;
	@Inject
	private changeLogsRepository: BookingChangeLogsRepository;
	@Inject
	private userContext: UserContext;

	private mapBookingState(booking: Booking): BookingJsonSchemaV1 {
		if (!booking)
			return {} as BookingJsonSchemaV1;

		const jsonObj = {
			id: booking.id,
			status: booking.status,
			startDateTime: booking.startDateTime,
			endDateTime: booking.startDateTime,
			serviceId: booking.serviceId,
			serviceName: booking.service.name,
			CitizenUinFin: booking.citizenUinFin,
			// TODO: ADD Citizen data;
			// CitizenName: string,
			// CitizenEmail: string,
			// CitizenPhone: string,
			// Location: string,
			// Description: string,
		} as BookingJsonSchemaV1;

		if (booking.serviceProviderId) {
			const serviceProvider = booking.serviceProvider;
			jsonObj.serviceProviderId = booking.serviceProviderId;
			jsonObj.serviceProviderName = serviceProvider.name;
			jsonObj.serviceProviderEmail = serviceProvider.email;
			// TODO: ADD SP PHONE jsonObj.serviceProviderPhone = serviceProvider.phone;
		}

		return jsonObj;
	}

	public async executeAndLogAction(booking: Booking, actionFunction: BookingActionFunction): Promise<Booking> {
		const user = await this.userContext.getCurrentUser();

		return await this.transactionManager.runInTransaction(BookingChangeLogsService.isolationLevel, async () => {
			const previousState = this.mapBookingState(booking);
			const [action, newBooking] = await actionFunction(booking);
			const newState = this.mapBookingState(newBooking);

			const changelog = BookingChangeLog.create({
				action,
				booking: newBooking,
				user,
				previousState,
				newState
			});

			await this.changeLogsRepository.saveLog(changelog);
			return newBooking;
		});
	}
}
