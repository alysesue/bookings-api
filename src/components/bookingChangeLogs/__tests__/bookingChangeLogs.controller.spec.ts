import { Container } from 'typescript-ioc';
import { BookingChangeLogsService } from '../bookingChangeLogs.service';
import { BookingChangeLogsController } from '../bookingChangeLogs.controller';
import { BookingChangeLog, ChangeLogAction, ServiceProvider, User } from '../../../models';
import { BookingBuilder } from '../../../models/entities/booking';
import { groupByKey } from '../../../tools/collections';

afterAll(() => {
	jest.resetAllMocks();
	if (global.gc) global.gc();
});

jest.mock('mol-lib-common', () => {
	const actual = jest.requireActual('mol-lib-common');
	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	const mock = (config: any) => {
		return (target: any, propertyKey: string | symbol, descriptor: PropertyDescriptor) => descriptor;
	};
	return {
		...actual,
		MOLAuth: mock,
	};
});

describe('BookingChangeLogs controller', () => {
	beforeAll(() => {
		Container.bind(BookingChangeLogsService).to(BookingChangeLogsServiceMock);
	});

	const adminMock = User.createAdminUser({
		molAdminId: 'd080f6ed-3b47-478a-a6c6-dfb5608a199d',
		userName: 'UserName',
		email: 'test@email.com',
		name: 'Name',
		agencyUserId: 'ABC123',
	});

	const booking = new BookingBuilder()
		.withServiceId(1)
		.withStartDateTime(new Date(Date.UTC(2020, 0, 1, 14, 0)))
		.withEndDateTime(new Date(Date.UTC(2020, 0, 1, 15, 0)))
		.build();
	booking.id = 1;
	booking.serviceProvider = ServiceProvider.create('name', 1, 'ad@as.com', '800 120 7163');

	it('should get logs', async () => {
		const controller = Container.get(BookingChangeLogsController);
		const changedSince = new Date(Date.UTC(2020, 0, 1, 14, 0));
		const changedUntil = new Date(Date.UTC(2020, 0, 31, 14, 0));

		const log = BookingChangeLog.create({
			booking,
			user: adminMock,
			action: ChangeLogAction.Create,
			previousState: { citizenName: 'a', citizenEmail: 'b@email.com' },
			newState: { citizenName: 'c' },
		});
		log.timestamp = new Date(Date.UTC(2020, 0, 1, 14, 0));

		const dataLogs = [log];

		BookingChangeLogsServiceMock.getLogs.mockImplementation(() =>
			Promise.resolve(groupByKey(dataLogs, (e) => e.booking.id)),
		);

		const result = await controller.getChangeLogs(changedSince, changedUntil, [2, 3], 1);
		expect(result).toMatchSnapshot();
	});
});

class BookingChangeLogsServiceMock implements Partial<BookingChangeLogsService> {
	public static getLogs = jest.fn<Promise<Map<number, BookingChangeLog[]>>, any>();

	public async getLogs(...params): Promise<any> {
		return await BookingChangeLogsServiceMock.getLogs(...params);
	}
}
