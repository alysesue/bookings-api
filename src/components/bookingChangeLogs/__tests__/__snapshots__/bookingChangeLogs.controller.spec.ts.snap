// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingChangeLogs controller should get logs 1`] = `
Array [
  BookingChangeLogResponse {
    "bookingId": 1,
    "changeLogs": Array [
      ChangeLogEntryResponse {
        "action": "create",
        "changes": Object {
          "citizenName": "c",
        },
        "previousBooking": Object {
          "citizenEmail": "b@email.com",
          "citizenName": "a",
          "schemaVersion": 1,
        },
        "timestamp": 2020-01-01T14:00:00.000Z,
        "user": UserTypeResponse {
          "admin": AdminUserContract {
            "agencyUserId": "ABC123",
            "email": "test@email.com",
          },
          "userType": "admin",
        },
      },
    ],
  },
]
`;
